tests/Filter/ComplexFilterTest.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+/*
+ * This file is part of the Behat Gherkin Parser.
+ * (c) Konstantin Kudryashov <ever.zet@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Tests\Behat\Gherkin\Filter;
+
+use Behat\Gherkin\Filter\ComplexFilter;
+use Behat\Gherkin\Node\FeatureNode;
+use Behat\Gherkin\Node\ScenarioInterface;
+use Behat\Gherkin\Node\ScenarioNode;
+
+class ComplexFilterTest extends FilterTestCase
+{
+    public function testFilterFeature(): void
+    {
+        $scenarios = [new ScenarioNode('Scenario#1', [], [], '', 1)];
+        $feature = new FeatureNode('Feature#1', null, [], null, $scenarios, '', '', null, 1);
+        $nonFilteringFilter = new class extends ComplexFilter {
+            public function isScenarioMatch(FeatureNode $feature, ScenarioInterface $scenario): bool
+            {
+                return true;
+            }
+
+            public function isFeatureMatch(FeatureNode $feature): bool
+            {
+                return true;
+            }
+        };
+
+        $filtered = $nonFilteringFilter->filterFeature($feature);
+
+        $this->assertCount(1, $scenarios = $filtered->getScenarios());
+        $this->assertSame('Scenario#1', $scenarios[0]->getTitle());
```

Comment:
Thinking about it more, I'm still not sure I see the value of this test.

I could make it pass just by changing the `ComplexFilter` implementation to be a no-op (which should surely break something):

```
public function filterFeature(FeatureNode $feature) {
  return $feature;
}
```

If it's worth having a test at all, then I think it probably at least needs to make sure the method does something? Even just a second example with an anonymous class that returns `false` and proves we get an empty feature?

Reply:
Hey, I hit a bit of a snag.

Seems that the original `ComplexFilter` never invoked abstract method `isFeatureMatch()`. Funnily enough, the only implementation of this filter (`TagFilter`) also doesn't invoke `isFeatureMatch()`.

This doesn't make much sense to me, but it means there isn't anything to test regarding `isFeatureMatch()` with respect to `ComplexFilter`.

Agreed?

Reply:
I guess `isFeatureMatch()` is just present to satisfy the `FeatureFilterInterface`? I agree if we don't provide an actual implementation of that then there's no need to test it.

I'm still not actually sure why we're testing an abstract class at all - normally I would just test them through their concretions. I suppose at least it will tell us if we break BC for end-users that have extended it.

Reply:
We test that abstract class because it is part of our public API (end users can extend it, it is not an internal one)